system/nxcamera/nxcamera.c

```diff
@@ -0,0 +1,889 @@
+ *   to perform the create for proper reference counting.
+ *
+ * Input Parameters:  None
+ *
+ * Returned values:
+ *   Pointer to the created context or NULL if there was an error.
+ *
+ ****************************************************************************/
+
+FAR struct nxcamera_s *nxcamera_create(void)
+{
+  FAR struct nxcamera_s *pcam;
+  int err;
+
+  /* Allocate the memory */
+
+  pcam = (FAR struct nxcamera_s *)calloc(1, sizeof(struct nxcamera_s));
+  if (pcam == NULL)
+    {
+      return NULL;
+    }
+
+  /* Initialize the context data */
+
+  pcam->loopstate = NXCAMERA_STATE_IDLE;
+  pcam->display_fd = -1;
+  pcam->capture_fd = -1;
+  err = pthread_mutex_init(&pcam->mutex, NULL);
+  if (err)
+    {
+      verr("ERROR: pthread_mutex_init failed: %d\n", err);
+      free(pcam);
+      pcam = NULL;
+    }
+
+  return pcam;
+}
+
+/****************************************************************************
+ * Name: nxcamera_release
+ *
+ *   nxcamera_release() reduces the reference count by one and if it
+ *   reaches zero, frees the context.
+ *
+ * Input Parameters:
+ *   pcam    Pointer to the NxCamera context
+ *
+ * Returned values:    None
+ *
+ ****************************************************************************/
+
+void nxcamera_release(FAR struct nxcamera_s *pcam)
+{
+  int      refcount;
+  FAR void *value;
+  int errcode;
+
+  /* Lock the mutex */
+
+  while ((errcode = pthread_mutex_trylock(&pcam->mutex)))
+    {
+      DEBUGASSERT(errcode > 0);
+
+      if (errcode != EBUSY)
+        {
+          verr("ERROR: pthread_mutex_trylock failed: %d\n", errcode);
+          return;
+        }
+    }
+
+  /* Check if there was a previous thread and join it if there was */
+
+  if (pcam->loop_id != 0)
+    {
+      pthread_mutex_unlock(&pcam->mutex);
+      pthread_join(pcam->loop_id, &value);
+      pcam->loop_id = 0;
+
+      while ((errcode = pthread_mutex_trylock(&pcam->mutex)))
+        {
+          DEBUGASSERT(errcode > 0);
+
+          if (errcode != EBUSY)
+            {
+              verr("ERROR: pthread_mutex_trylock failed: %d\n", errcode);
+              return;
+            }
+        }
+    }
+
+  /* Reduce the reference count */
+
+  refcount = pcam->crefs--;
+  pthread_mutex_unlock(&pcam->mutex);
+
+  /* If the ref count *was* one, then free the context */
+
+  if (refcount == 1)
+    {
+      pthread_mutex_destroy(&pcam->mutex);
+      free(pcam);
+    }
+}
+
+/****************************************************************************
+ * Name: nxcamera_reference
+ *
+ *   nxcamera_reference() increments the reference count by one.
+ *
+ * Input Parameters:
+ *   pcam    Pointer to the NxCamera context
+ *
+ * Returned values:    None
+ *
+ ****************************************************************************/
+
+void nxcamera_reference(FAR struct nxcamera_s *pcam)
+{
+  int errcode;
+
+  /* Lock the mutex */
+
+  while ((errcode = pthread_mutex_trylock(&pcam->mutex)))
```

Comment:
What is the benefit of `while` + `pthread_mutex_trylock` vs `pthread_mutex_lock` here?